# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
inventories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB


In [4]:
inventory_sets.head()

,inventory_id,set_num,quantity
0,35,75911-1,1
1,35,75912-1,1
2,39,75048-1,1
3,39,75053-1,1
4,50,4515-1,1


In [5]:
sets.head()

,set_num,name,year,theme_id,num_parts
0,001-1,Gears,1965,1,43
1,0011-2,Town Mini-Figures,1979,67,12
2,0011-3,Castle 2 for 1 Bonus Offer,1987,199,0
3,0012-1,Space Mini-Figures,1979,143,12
4,0013-1,Space Mini-Figures,1979,143,12


In [6]:
themes.head()

,id,name,parent_id
0,1,Technic,NaN
1,3,Competition,1.0
2,4,Expert Builder,1.0
3,16,RoboRiders,1.0
4,17,Speed Slammers,1.0


In [7]:
inventory_minifigs.head()

,inventory_id,fig_num,quantity
0,3,fig-001549,1
1,4,fig-000764,1
2,19,fig-000555,1
3,25,fig-000574,1
4,26,fig-000842,1


In [8]:
minifigs.head()

,fig_num,name,num_parts
0,fig-000001,Toy Store Employee,4
1,fig-000002,Customer Kid,4
2,fig-000003,"Assassin Droid, White",8
3,fig-000004,Basic Figure,4
4,fig-000005,Captain America with Short Legs,3


In [9]:
inventory_parts.head()

,inventory_id,part_num,color_id,quantity,is_spare
0,1,48379c01,72,1,f
1,1,48395,7,1,f
2,1,stickerupn0077,9999,1,f
3,1,upn0342,0,1,f
4,1,upn0350,25,1,f


In [10]:
part_categories.head()

,id,name
0,1,Baseplates
1,3,Bricks Sloped
2,4,"Duplo, Quatro and Primo"
3,5,Bricks Special
4,6,Bricks Wedged


In [11]:
parts.head()

,part_num,name,part_cat_id,part_material
0,003381,Sticker Sheet for Set 663-1,58,Plastic
1,003383,"Sticker Sheet for Sets 618-1, 628-2",58,Plastic
2,003402,"Sticker Sheet for Sets 310-3, 311-1, 312-3",58,Plastic
3,003429,Sticker Sheet for Set 1550-1,58,Plastic
4,003432,"Sticker Sheet for Sets 357-1, 355-1, 940-1",58,Plastic


In [12]:
colors.head()

,id,name,rgb,is_trans
0,-1,[Unknown],0033B2,f
1,0,Black,05131D,f
2,1,Blue,0055BF,f
3,2,Green,237841,f
4,3,Dark Turquoise,008F9B,f


In [13]:
part_relationships.head()

,rel_type,child_part_num,parent_part_num
0,P,87079pr9974,87079
1,P,3960pr9971,3960
2,R,98653pr0003,98086pr0003
3,R,98653pr0003,98088pat0003
4,R,98653pr0003,98089pat0003


In [14]:
elements.head()

,element_id,part_num,color_id
0,6300211,67906c01,14
1,4566309,2564,0
2,4275423,53657,1004
3,6194308,92926,71
4,6229123,26561,4


## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [15]:
def data_count(data):
    return data.shape[0]

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [16]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [17]:
Q2 = round(sets['num_parts'].mean())
Q2

161

### **Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [18]:
Q3 = round(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [19]:
Q4 = round(sets['num_parts'].max())
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [20]:
Q5 = sets[sets['num_parts'] == sets['num_parts'].max()]
Q5

,set_num,name,year,theme_id,num_parts
3867,31203-1,World Map,2021,709,11695


**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [21]:
def select_set_row(data):
    row = sets[sets['set_num'] == data]
    return row

#select_set_row('0011-3')

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [22]:
def select_set_numparts(data):
    rows = sets[sets['num_parts'] == data]
    rows.sort_values(by = ['year', 'set_num'], ascending = True, inplace = True)
    return rows

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [23]:
select_set_numparts(0)

C:\Users\coltm\AppData\Local\Temp\ipykernel_5512\125690697.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rows.sort_values(by = ['year', 'set_num'], ascending = True, inplace = True)


,set_num,name,year,theme_id,num_parts
11061,700.1-2,Extra-Large Gift Set (Mursten),1953,366,0
17563,b55dk-01,Lego Mursten - System i Leg Byggebog,1955,497,0
17564,b56de-01,System im Spiel,1956,497,0
11072,700.2-2,Gift Package,1957,366,0
11078,700.3A-2,Gift Package,1957,372,0
...,...,...,...,...,...
15619,854195-1,Ernie Key Chain,2022,503,0
15620,854197-1,Demogorgon Key Chain,2022,503,0
17170,9780744054583-1,DC Comics Super Heroes: Character Encyclopedia...,2022,497,0
17193,9780794447557-1,City: Color The City,2022,497,0


**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [24]:
Q8A = sets['year'].min()
Q8A

1949

In [25]:
Q8B = sets['year'].max()
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [26]:
def select_set_year(year):
    rows = sets[sets['year'] == year] 
    rows.sort_values(by = 'set_num', ascending = True, inplace = True)
    return rows

#select_set_year(1983)

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [27]:
def theme_by_year(year):
    sets1 = sets.copy()
    themes1 = themes.copy()
    sets1.rename(columns={'theme_id': 'id'}, inplace = True)
    themes1.rename(columns={'name':'name_themes'}, inplace = True)
    sets_and_themes = pd.merge(sets1, themes1, on = 'id')
    data = sets_and_themes.loc[sets_and_themes['year'] == year, ['id', 'name_themes']]
    data.drop_duplicates(['id'], inplace = True)
    data.sort_values(by=['id'], ascending=True, inplace=True)
    data = data.reset_index(drop=True)
    return data

In [28]:
#theme_by_year(2022)

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [29]:
theme_by_year(1960)

,id,name_themes
0,371,Supplemental
1,497,Books
2,513,Classic


**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [30]:
def theme_by_name(themename):
    sets2 = sets.copy()
    themes2 = themes.copy()
    sets2.rename(columns = {'id':'theme_id'}, inplace = True)
    sets2.rename(columns = {'name': 'name_sets'}, inplace = True) #change 'name' in sets to name_sets
    themes2.rename(columns = {'id':'theme_id'}, inplace = True) #change 'id' in themes to theme_id
    themes2.rename(columns = {'name':'name_themes'}, inplace = True) #change 'name'in themes to name_themes
    sets_and_themes = pd.merge(sets2, themes2, on = 'theme_id') #merge sets and themes over the theme_id column
    data = sets_and_themes.loc[sets_and_themes['name_themes'] == themename, 
                               ['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']]
    #data.drop_duplicates(['name_themes'], inplace = True) #eliminated all but first occurance of "Gears", so needed to take out from last function
    data.sort_values(by=['set_num'], ascending=True, inplace=True)
    data = data.reset_index(drop=True)
    return data

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [31]:
theme_by_name('Gear')

,set_num,name_sets,year,num_parts,theme_id,name_themes
0,11138-1,Bionicle Barraki Wallet,2008,0,501,Gear
1,119645111436-1,Classic Space Astronaut Plush (Red),2021,0,501,Gear
2,119645111504-1,Classic Space Astronaut Plush (Blue),2021,0,501,Gear
3,119645111740-1,Classic Space Astronaut Plush (White),2021,0,501,Gear
4,119645111818-1,Classic Space Astronaut Plush (Pink),2021,0,501,Gear
...,...,...,...,...,...,...
1899,WATCH-3,Watch System Voyager (UFO Space),1997,29,501,Gear
1900,WIIGAME-1,Bionicle Heroes - Wii,2007,0,501,Gear
1901,WIIGAME-2,Indiana Jones: The Original Adventures Video G...,2008,0,501,Gear
1902,WOODENELEPHANT-1,Original Wooden Elephant Pull-Along Toy and Print,2021,0,501,Gear


**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [32]:
def theme_by_setnum(setnum):
    sets3 = sets.copy()
    sets3.rename(columns = {'theme_id':'id'}, inplace = True)
    themes3 = themes.copy()
    themes3.rename(columns = {'name':'theme_name'}, inplace = True)
    sets_and_themes = pd.merge(sets3, themes3, on = 'id') #merge sets and themes over the id column
    data = sets_and_themes.loc[sets_and_themes['set_num'] == setnum,['theme_name']].values
    #change to string
    data_string = str(data)
    data_string = data_string[3:-3]
    return data_string

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [33]:
theme_by_setnum('75810-1')

'Stranger Things'

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [34]:
MillenniumFalcon = sets[sets['name'] == 'Millennium Falcon']
MillenniumFalcon.sort_values(by=['num_parts'], ascending=False)

,set_num,name,year,theme_id,num_parts
12884,75192-1,Millennium Falcon,2017,171,7541
209,10179-1,Millennium Falcon,2007,171,5197
12959,75257-1,Millennium Falcon,2019,158,1351
12790,75105-1,Millennium Falcon,2015,158,1330
13846,7965-1,Millennium Falcon,2011,158,1254
12450,7190-1,Millennium Falcon,2000,158,671
12719,75030-1,Millennium Falcon,2014,158,94
6570,4488-1,Millennium Falcon,2003,158,87
18488,TRUFALCON-1,Millennium Falcon,2016,158,44
16458,911607-1,Millennium Falcon,2016,158,42


**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [35]:
inventories.head(1)

,id,version,set_num
0,1,1,7922-1


In [36]:
sets.head(1)

,set_num,name,year,theme_id,num_parts
0,001-1,Gears,1965,1,43


In [37]:
inventory_minifigs.head(1)

,inventory_id,fig_num,quantity
0,3,fig-001549,1


In [38]:
minifigs.head(1)

,fig_num,name,num_parts
0,fig-000001,Toy Store Employee,4


In [39]:
def minfigcal(setnum):
    inventories4 = inventories.copy()
    inventory_minifigs4 = inventory_minifigs.copy()
    minifigs4 = minifigs.copy()
    sets4 = sets.copy()
    inventory_minifigs4.rename(columns={'inventory_id': 'id'}, inplace = True)
    #combine sets, Inventories, inventory_minfigs, minifigs
    minfig_df = pd.merge(inventories4, sets4, on = 'set_num')
    minfig_df = pd.merge(minfig_df, inventory_minifigs4, on = 'id')
    minfig_df = pd.merge(minfig_df, minifigs4, on = 'fig_num')
    data = minfig_df.loc[minfig_df['set_num'] == setnum, ['quantity']]
    return len(data)

Q13 = minfigcal('75192-1')
Q13

8

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [40]:
def get_minifigs(setnum, version):
    inventories4 = inventories.copy()
    inventory_minifigs4 = inventory_minifigs.copy()
    minifigs4 = minifigs.copy()
    sets4 = sets.copy()
    sets4.rename(columns={'name':'name_sets'}, inplace = True)
    inventory_minifigs4.rename(columns={'inventory_id': 'id'}, inplace = True)
    minifigs4.rename(columns={'name':'fig_name', 'num_parts':'fig_num_parts'}, inplace = True)
    #combine sets, Inventories, inventory_minfigs, minifigs
    minfig_df = pd.merge(inventories4, sets4, on = 'set_num')
    minfig_df = pd.merge(minfig_df, inventory_minifigs4, on = 'id')
    minfig_df = pd.merge(minfig_df, minifigs4, on = 'fig_num')
    #Refine data set
    data = minfig_df.loc[(minfig_df['set_num'] == setnum) & (minfig_df['version'] == version),['set_num', 'fig_num','fig_name', 'fig_num_parts']]
    data = data.sort_values(by=['fig_num'], ascending=True)
    data = data.reset_index(drop=True)
    return data

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [41]:
Q15 = get_minifigs('75192-1', 1)
Q15

,set_num,fig_num,fig_name,fig_num_parts
0,75192-1,fig-001714,"Chewbacca, Dark Brown",3
1,75192-1,fig-001810,"Finn, Black Undershirt",4
2,75192-1,fig-002057,"Rey, Dark Tan Robe",4
3,75192-1,fig-002514,"C-3PO, Pearl Gold, Colorful Wires, Printed Legs",3
4,75192-1,fig-002544,BB-8,2
5,75192-1,fig-002549,"Princess Leia, Hoth Outfit, White Jacket, Prin...",4
6,75192-1,fig-002550,"Han Solo, Dark Blue Jacket, Dark Brown Legs, P...",4
7,75192-1,fig-002551,"Han Solo, Old, Angry",4


**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [42]:
# Harry Potter's Diagon Alley Set 1
get_minifigs('10217-1', 1)

,set_num,fig_num,fig_name,fig_num_parts
0,10217-1,fig-002229,Skeleton (Standard) (3626b Head),6
1,10217-1,fig-006075,"Rubeus Hagrid, Dark Brown Overcoat",3
2,10217-1,fig-006084,"Lucius Malfoy, Black Robes and Cape, Death Eat...",5
3,10217-1,fig-006130,"Harry Potter, Open Dark Blue Jacket over White...",4
4,10217-1,fig-006138,Fenrir Greyback,4
5,10217-1,fig-006162,"Ron Weasley, Red Tartan Sweater Vest, Reddish ...",4
6,10217-1,fig-006163,Fred / George Weasley in Dark Orange Suit,4
7,10217-1,fig-006164,Garrick Ollivander,4
8,10217-1,fig-006165,"Goblin, Black Jacket and Light Bluish Gray Legs",3
9,10217-1,fig-006166,"Goblin, Black Jacket and Reddish Brown Legs",3


In [43]:
# Harry Potter's Diagon Alley Set 2
get_minifigs('10217-1', 2)

,set_num,fig_num,fig_name,fig_num_parts
0,10217-1,fig-002229,Skeleton (Standard) (3626b Head),6
1,10217-1,fig-006075,"Rubeus Hagrid, Dark Brown Overcoat",3
2,10217-1,fig-006084,"Lucius Malfoy, Black Robes and Cape, Death Eat...",5
3,10217-1,fig-006130,"Harry Potter, Open Dark Blue Jacket over White...",4
4,10217-1,fig-006138,Fenrir Greyback,4
5,10217-1,fig-006162,"Ron Weasley, Red Tartan Sweater Vest, Reddish ...",4
6,10217-1,fig-006163,Fred / George Weasley in Dark Orange Suit,4
7,10217-1,fig-006164,Garrick Ollivander,4
8,10217-1,fig-006165,"Goblin, Black Jacket and Light Bluish Gray Legs",3
9,10217-1,fig-006166,"Goblin, Black Jacket and Reddish Brown Legs",3


**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [44]:
def minifigs_from_themes(themeid):
    #copys of original data sets to not change the original sets and affect code before this
    inventories4 = inventories.copy()
    inventory_minifigs4 = inventory_minifigs.copy()
    minifigs4 = minifigs.copy()
    sets4 = sets.copy()
    sets4.rename(columns={'name':'name_sets'}, inplace = True)
    inventory_minifigs4.rename(columns={'inventory_id': 'id'}, inplace = True)
    minifigs4.rename(columns={'name':'fig_name', 'num_parts':'fig_num_parts'}, inplace = True)
    #combine sets, Inventories, inventory_minfigs, minifigs
    minfig_df = pd.merge(inventories4, sets4, on = 'set_num')
    minfig_df = pd.merge(minfig_df, inventory_minifigs4, on = 'id')
    minfig_df = pd.merge(minfig_df, minifigs4, on = 'fig_num')
    #refine the data
    data = minfig_df.loc[minfig_df['theme_id'] == themeid, ['fig_num', 'fig_name']]
    data.sort_values(by=['fig_num'], ascending=False)
    data = data.reset_index(drop=True)
    return data

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [45]:
minifigs_from_themes(171)

,fig_num,fig_name
0,fig-002524,"Stormtrooper, Dark Blue Helmet Vents"
1,fig-002524,"Stormtrooper, Dark Blue Helmet Vents"
2,fig-004083,"Bespin Guard, Jacket with Gold Trim"
3,fig-004084,"Han Solo, White Shirt, Reddish Brown Legs, Dua..."
4,fig-004085,"Boba Fett, Helmet, Jetpack Neckwear, Dark Tan ..."
...,...,...
183,fig-011942,"AT-AT Driver - Light Bluish Grey Helmet, Sand ..."
184,fig-011944,"Snowtrooper, White Hips, Reddish Brown Head, F..."
185,fig-011945,"Snowtrooper, White Hips, Light Flesh Head, Female"
186,fig-011946,"Snowtrooper, White Hips, Light Flesh Head, Off..."


**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [46]:
def sets_from_minifig(minifignum):
    #copys of original data sets to not change the original sets and affect code before this
    inventories4 = inventories.copy()
    inventory_minifigs4 = inventory_minifigs.copy()
    minifigs4 = minifigs.copy()
    sets4 = sets.copy()
    sets4.rename(columns={'name':'name_sets'}, inplace = True)
    inventory_minifigs4.rename(columns={'inventory_id': 'id'}, inplace = True)
    minifigs4.rename(columns={'name':'fig_name', 'num_parts':'fig_num_parts'}, inplace = True)
    #combine sets, Inventories, inventory_minfigs, minifigs
    minfig_df = pd.merge(inventories4, sets4, on = 'set_num')
    minfig_df = pd.merge(minfig_df, inventory_minifigs4, on = 'id')
    minfig_df = pd.merge(minfig_df, minifigs4, on = 'fig_num')
    #refine the data
    data = minfig_df.loc[minfig_df['fig_num'] == minifignum,['set_num','version','name_sets','fig_num', 'fig_name']]
    data = data.sort_values(by=['set_num', 'version'], ascending=True)
    data = data.reset_index(drop=True)
    return data

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [47]:
sets_from_minifig('fig-001039')

,set_num,version,name_sets,fig_num,fig_name
0,30100-1,1,Andrea on the Beach,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."
1,30107-1,1,Birthday Party,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."
2,3061-1,1,City Park Café,fig-001039,"Andrea - Bright Light Orange Top, Light Aqua S..."


**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [48]:
themes5 = themes.copy()
sets5 = sets.copy()
inventories5 = inventories.copy()
inventory_parts5 = inventory_parts.copy()
parts5 = parts.copy()
colors5 = colors.copy()
    # changing column names for alignment and required columns
parts5.rename(columns = {'name':'name_part'}, inplace = True) ## Parts - Name to name_part
colors5.rename(columns = {'name':'name_color', 'id':'color_id'}, inplace = True) ## Colors - Name to name_color & id to color_id
inventory_parts5.rename(columns = {'inventory_id':'id'}, inplace = True)
themes5.rename(columns = {'id':'theme_id', 'name':'name_themes'}, inplace = True)
sets5.rename(columns = {'name':'name_sets'}, inplace = True)   #change columns needed for merge alignment
    # merge dataframes together
mass_df = pd.merge(inventory_parts5, parts5, on = 'part_num')
mass_df = pd.merge(mass_df, colors5, on = 'color_id')
mass_df = pd.merge(mass_df, inventories5, on = 'id')
mass_df = pd.merge(mass_df, sets5, on = 'set_num')
mass_df = pd.merge(mass_df, themes5, on = 'theme_id')
mass_df.head()

,id,part_num,color_id,quantity,is_spare,name_part,part_cat_id,part_material,name_color,rgb,is_trans,version,set_num,name_sets,year,theme_id,num_parts,name_themes,parent_id
0,1,48379c01,72,1,f,"Large Figure Torso and Legs, Promo Figure Base...",41,Plastic,Dark Bluish Gray,6C6E68,f,1,7922-1,McDonald's Sports Set Number 6 - Orange Vest S...,2004,460,5,Gravity Games,458.0
1,1,upn0342,0,1,f,Sports Promo Paddle from McDonald's Sports Sets,27,Plastic,Black,05131D,f,1,7922-1,McDonald's Sports Set Number 6 - Orange Vest S...,2004,460,5,Gravity Games,458.0
2,1,48395,7,1,f,Sports Snowboard from McDonald's Promotional Set,27,Plastic,Light Gray,9BA19D,f,1,7922-1,McDonald's Sports Set Number 6 - Orange Vest S...,2004,460,5,Gravity Games,458.0
3,1,stickerupn0077,9999,1,f,Sticker Sheet for Set 7922-1,58,Plastic,[No Color/Any Color],05131D,f,1,7922-1,McDonald's Sports Set Number 6 - Orange Vest S...,2004,460,5,Gravity Games,458.0
4,1,upn0350,25,1,f,Sports Promo Figure Head Torso Assembly McDona...,13,Plastic,Orange,FE8A18,f,1,7922-1,McDonald's Sports Set Number 6 - Orange Vest S...,2004,460,5,Gravity Games,458.0


In [49]:
def set_parts(setnum, version):
    data = mass_df.loc[(mass_df['set_num'] == setnum) & (mass_df['version'] == version), ['set_num','version','name_sets','year','name_themes','part_num','name_part','name_color', 'quantity', 'is_spare']]
    data.sort_values(by = ['part_num', 'name_color'], ascending=True, inplace=True)
    data = data.reset_index(drop=True)
    return data

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [50]:
set_parts(75192-1, 1)

,set_num,version,name_sets,year,name_themes,part_num,name_part,name_color,quantity,is_spare


## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
